### ISSUE TITLE
When running quickstart.ipynb, loading model in int8 and fp16 occupy significantly different amounts of GPU memory.

### ISSUE THREAD

**Issue by lankuohsing** [meta-llama/llama-recipes#374]
I am trying finetuning llama2 7B with lora by running quickstart.ipynb(https://github.com/facebookresearch/llama-recipes/blob/main/examples/quickstart.ipynb), using an A100 40G GPU.
When I load the model in int8 and create a PeftModel in int8 (just as the original setting in quickstart.ipynb), the training occupies 14 GB GPU memory (batch_size is set to be 2). 
However, when I load the model in fp16 and create a PeftModel in fp16,  the training occupies 40 GB GPU memory (batch_size is set to be 1 ). 

The part of the code I modified is shown below:
```
model =LlamaForCausalLM.from_pretrained(model_id, device_map='auto', torch_dtype=torch.float16)
...
config = {
    'lora_config': lora_config,
    'learning_rate': 1e-5,# from 1e-4 to 1e-5
    'num_train_epochs': 1,
    'gradient_accumulation_steps': 4,#from 2 to 4
    'per_device_train_batch_size': 1,#from 2 to 1
    'gradient_checkpointing': False,
}
...
# model = prepare_model_for_int8_training(model)
model = get_peft_model(model, peft_config)
...
```

Can someone explain why there is such a huge difference of GPU memory consumption?

**Comment by HamidShojanazeri** [comment]
@lankuohsing reading the bit&bytes [docs](https://huggingface.co/docs/bitsandbytes/main/en/optimizers) might be helpful, this support a bunch of functionality beside int8 matrix multiplication, it also supports int8 optimizer that can significanly reduce the memory requirements and if you look at [prepare_model_for_int8_train](https://github.com/huggingface/peft/blob/main/src/peft/utils/other.py#L143) it also enables gradient checkpointing which is big memory saver. That should sort of explain the difference.

**Comment by wukaixingxp** [comment]
Hi! It seems that a solution has been provided to the issue and there has not been a follow-up conversation for a long time. I will close this issue for now and feel free to reopen it if you have any questions!
